Error while checking build/O2/fullCI for 0812196ed8f87e4c740a28b85edd9b51eedbfda7 at 2024-01-19 08:46:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
[FATAL] error while setting up workflow in /sw/BUILD/18aa623c4b3ef92cc884739b95ff892268cdc5f9/O2Physics/stage/bin/o2-analysis-je-emcal-correction-task: Geometry not initialized
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12560/0812196ed8f87e4c740a28b85edd9b51eedbfda7/build_O2_fullCI/pretty.html).

--- a/Detectors/EMCAL/base/include/EMCALBase/ClusterFactory.h
+++ b/Detectors/EMCAL/base/include/EMCALBase/ClusterFactory.h
@@ -188,7 +188,7 @@ class ClusterFactory
 
   ///
   /// Dummy constructor
-  ClusterFactory();
+  ClusterFactory() = default;
 
   ///
   /// \brief Constructor initializing the ClusterFactory

--- a/Detectors/EMCAL/base/include/EMCALBase/Geometry.h
+++ b/Detectors/EMCAL/base/include/EMCALBase/Geometry.h
@@ -70,6 +70,7 @@ class Geometry
 
   /// \brief Get geometry instance. It should have been set before.
   /// \return the pointer of the unique instance of the geometry
+  /// \throw GeometryNotInitializedException in case the geometry is not initialized
   static Geometry* GetInstance();
 
   /// \brief Get instance of the EMCAL geometry

--- a/Detectors/EMCAL/base/include/EMCALBase/GeometryBase.h
+++ b/Detectors/EMCAL/base/include/EMCALBase/GeometryBase.h
@@ -33,6 +33,23 @@ enum AcceptanceType_t { EMCAL_ACCEPTANCE = 1,
 
 const std::string DEFAULT_GEOMETRY = "EMCAL_COMPLETE12SMV1_DCAL_8SM";
 
+/// \class GeometryNotInitializedException
+/// \brief Error handling access to non-initialized geometry
+/// \ingroup EMCALbase
+class GeometryNotInitializedException final : public std::exception
+{
+ public:
+  /// \brief Constructor
+  GeometryNotInitializedException() = default;
+
+  /// \brief Destructor
+  ~GeometryNotInitializedException() noexcept final = default;
+
+  /// \brief Access to error message
+  /// \return Error message
+  const char* what() const noexcept { return "Geometry not initialized"; }
+};
+
 /// \class InvalidModuleException
 /// \brief Error Handling when an invalid module ID (outside the limits) is called
 /// \ingroup EMCALbase

--- a/Detectors/EMCAL/base/src/ClusterFactory.cxx
+++ b/Detectors/EMCAL/base/src/ClusterFactory.cxx
@@ -25,17 +25,9 @@
 
 using namespace o2::emcal;
 
-template <typename InputType>
-ClusterFactory<InputType>::ClusterFactory()
-{
-  mGeomPtr = o2::emcal::Geometry::GetInstance();
-}
-
 template <class InputType>
 ClusterFactory<InputType>::ClusterFactory(gsl::span<const o2::emcal::Cluster> clustersContainer, gsl::span<const InputType> inputsContainer, gsl::span<const int> cellsIndices)
 {
-  mGeomPtr = o2::emcal::Geometry::GetInstance();
-
   setContainer(clustersContainer, inputsContainer, cellsIndices);
 }
 

--- a/Detectors/EMCAL/base/src/Geometry.cxx
+++ b/Detectors/EMCAL/base/src/Geometry.cxx
@@ -190,6 +190,9 @@ Geometry::~Geometry()
 Geometry* Geometry::GetInstance()
 {
   Geometry* rv = static_cast<Geometry*>(sGeom);
+  if (!rv) {
+    throw GeometryNotInitializedException();
+  }
   return rv;
 }
 

--- a/Detectors/EMCAL/calib/src/BadChannelMap.cxx
+++ b/Detectors/EMCAL/calib/src/BadChannelMap.cxx
@@ -72,25 +72,24 @@ TH2* BadChannelMap::getHistogramRepresentation() const
             MAXCOLS = 96;
   auto hist = new TH2S("badchannelmap", "Bad Channel Map", MAXCOLS, -0.5, double(MAXCOLS) - 0.5, MAXROWS, -0.5, double(MAXROWS) - 0.5);
   hist->SetDirectory(nullptr);
-  auto geo = Geometry::GetInstance();
-  if (!geo) {
+  try {
+    auto geo = Geometry::GetInstance();
+    for (size_t cellID = 0; cellID < mBadCells.size(); cellID++) {
+      int value(0);
+      if (mBadCells.test(cellID)) {
+        value = 1;
+      } else if (mDeadCells.test(cellID)) {
+        value = 2;
+      } else if (mWarmCells.test(cellID)) {
+        value = 3;
+      }
+      if (value) {
+        auto position = geo->GlobalRowColFromIndex(cellID);
+        hist->Fill(std::get<1>(position), std::get<0>(position), value);
+      }
+    }
+  } catch (o2::emcal::GeometryNotInitializedException& e) {
     LOG(error) << "Geometry needs to be initialized";
-    return hist;
-  }
-
-  for (size_t cellID = 0; cellID < mBadCells.size(); cellID++) {
-    int value(0);
-    if (mBadCells.test(cellID)) {
-      value = 1;
-    } else if (mDeadCells.test(cellID)) {
-      value = 2;
-    } else if (mWarmCells.test(cellID)) {
-      value = 3;
-    }
-    if (value) {
-      auto position = geo->GlobalRowColFromIndex(cellID);
-      hist->Fill(std::get<1>(position), std::get<0>(position), value);
-    }
   }
   return hist;
 }

--- a/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h
+++ b/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h
@@ -63,11 +63,11 @@ class EMCALCalibExtractor
   EMCALCalibExtractor()
   {
     LOG(info) << "initialized EMCALCalibExtractor";
-    if (!mGeometry) {
+    try {
+      // Try to access geometry initialized ountside
       mGeometry = o2::emcal::Geometry::GetInstance();
-      if (!mGeometry) {
-        mGeometry = o2::emcal::Geometry::GetInstanceFromRunNumber(300000); // fallback option
-      }
+    } catch (o2::emcal::GeometryNotInitializedException& e) {
+      mGeometry = o2::emcal::Geometry::GetInstanceFromRunNumber(300000); // fallback option
     }
     // mNcells = mGeometry->GetNCells();
   };

--- a/Detectors/EMCAL/calibration/src/EMCALCalibExtractor.cxx
+++ b/Detectors/EMCAL/calibration/src/EMCALCalibExtractor.cxx
@@ -30,7 +30,6 @@ boostHisto EMCALCalibExtractor::buildHitAndEnergyMeanScaled(double emin, double 
 {
   // create the output histogram
   boostHisto eSumHistoScaled = boost::histogram::make_histogram(boost::histogram::axis::regular<>(100, 0, 100, "t-texp"), boost::histogram::axis::integer<>(0, mNcells, "CELL ID"));
-
   // create a slice for each cell with energies ranging from emin to emax
   auto hEnergyCol = boost::histogram::make_histogram(boost::histogram::axis::regular<>(100, 0, 100., "t-texp"));
   auto hEnergyRow = boost::histogram::make_histogram(boost::histogram::axis::regular<>(250, 0, 250., "t-texp"));
@@ -47,9 +46,8 @@ boostHisto EMCALCalibExtractor::buildHitAndEnergyMeanScaled(double emin, double 
 
     for (int cellID = 0; cellID < mNcells; cellID++) {
       auto tempSlice = boost::histogram::algorithm::reduce(cellAmplitude, boost::histogram::algorithm::shrink(cellID, cellID), boost::histogram::algorithm::shrink(emin, emax));
-      auto geo = Geometry::GetInstance();
       // (0 - row, 1 - column)
-      auto position = geo->GlobalRowColFromIndex(cellID);
+      auto position = mGeometry->GlobalRowColFromIndex(cellID);
       int row = std::get<0>(position);
       int col = std::get<1>(position);
 
@@ -101,9 +99,8 @@ boostHisto EMCALCalibExtractor::buildHitAndEnergyMeanScaled(double emin, double 
 
     // Scale each cell by the deviation of the mean of the column and the global mean
     for (int iCell = 0; iCell < mNcells; iCell++) {
-      auto geo = Geometry::GetInstance();
       // (0 - row, 1 - column)
-      auto position = geo->GlobalRowColFromIndex(iCell);
+      auto position = mGeometry->GlobalRowColFromIndex(iCell);
       int col = std::get<1>(position);
       if (hEnergyCol.at(col) > 0.) {
         // will need to change the 100 depending on the number of energy bins we end up having
@@ -115,9 +112,8 @@ boostHisto EMCALCalibExtractor::buildHitAndEnergyMeanScaled(double emin, double 
 
     // Scale each cell by the deviation of the mean of the row and the global mean
     for (int iCell = 0; iCell < mNcells; iCell++) {
-      auto geo = Geometry::GetInstance();
       // (0 - row, 1 - column)
-      auto position = geo->GlobalRowColFromIndex(iCell);
+      auto position = mGeometry->GlobalRowColFromIndex(iCell);
       int row = std::get<0>(position);
       if (hEnergyRow.at(row) > 0.) {
         // will need to change the 100 depending on the number of energy bins we end up having

